base/inference.jl

```diff
@@ -3405,13 +3408,20 @@ function effect_free(e::ANY, src::CodeInfo, mod::Module, allow_volatile::Bool)
                             # first argument must be immutable to ensure e is affect_free
                             a = ea[2]
                             typ = widenconst(exprtype(a, src, mod))
-                            if !isa(typ, DataType) || typ.mutable || typ.abstract
+                            if !isa(typ, DataType) || typ.mutable || (typ.abstract && !isconstType(typ))
```

Comment:
Precompile could. But I have trouble believing we should add that bug. Can you actually observe the benefit of this change for well-written code?